Add mqtt client

eporsche opened the pull request:
Not sure if you want to add this to the main repo. But this will add a mqtt client to listen for commands.

Create a config_local.py as in the steps to setup a local dev environment.
Add some more variables:

MQTT = "True"
BROKER_ADDRESS=""
BROKER_PORT="1883"
BROKER_USERNAME=""
BROKER_PASSWORD=""
MQTT_TOPIC="hasip/execute"

Build the local.dockerfile and run the container with --net=host.
Now send the json files to control sip to the MQTT Topic.

arnonym committed:
Update pjsip to latest version.

arnonym committed:
Changes for next repository.

Commit:
adding mqtt wrapper

Commit:
do not add optional parameter to concfig.py

arnonym commented:
That's a cool idea! To merge that into the "offical" ha-sip I would like to unify those approaches a bit more. I think adding an official COMMAND_SOURCE option to the ha-sip configuration(config.json) with values "stdin" and "mqtt" and also adding BROKER_ADDRESS, BROKER_PORT, BROKER_USERNAME, BROKER_PASSWORD and MQTT_TOPIC there would accomplish this for the config part and you could run the mqtt mode inside home-assistant OS.

I've just pushed a version to the next branch which takes the configuration from environment variables instead of replacing the values in the python script (currently only tested locally with a .env file). With that in place you should not need a special docker image, but could use the official one and feed the environment variables externally or through `docker compose`.

What do you think? Did I miss anything?

arnonym committed:
Use environment variables to configure ha-sip internally.

arnonym committed:
Use environment variables to configure ha-sip internally.

arnonym committed:
Use environment variables to configure ha-sip internally.

arnonym committed:
Use environment variables to configure ha-sip internally.

arnonym committed:
Use environment variables to configure ha-sip internally.

arnonym committed:
Changes for next repository.

arnonym commented:
I was now able to run ha-sip through `docker compose` with a .env file for the configuration. I thought about putting the MQTT stuff in config.json again, and for now I would not do that, because there is probably no use-case, and it will only confuse home-assistant OS users.

I've also put COMMAND_SOURCE into config.py, and for MQTT it should be "mqtt" (as specified in the .env file).

Commit:
seperation mqtt from config

Commit:
further changes

Commit:
no need for local dockerfile

eporsche commented:
Nice - mqtt stuff is now being read from the .env file and is working. 

I changed the docker-compose.yml to build locally - since for mqtt to work the dependency paho-mqtt is required. 
Can be changed back when the official image contains that package.

I also added this dependency for the local dev environment.

arnonym commented on ha-sip/src/mqtt.py:
```diff
@@ -0,0 +1,33 @@
+        self.client.username_pw_set(self.username, self.password)
+
+    def on_connect(self, client, userdata, flags, reason_code, properties):
+        print(f"Connected with result code {reason_code}")
```
Please change prints to `log(None, ...)`

arnonym commented on ha-sip/src/main.py:
```diff
@@ -99,7 +105,22 @@ def main():
+
+    if mqtt_mode:
+        # Configuration
+        broker_address = os.environ.get('BROKER_ADDRESS','')
```
Can you please add the environment variables to .env.example?

arnonym commented on ha-sip/src/mqtt.py:
```diff
@@ -0,0 +1,33 @@
+from command_client import CommandClient
+
+class MQTTClient:
+    def __init__(self, broker_address, port, username, password, topic, command_handler):
```
Please add type annotations for the parameters

arnonym reviewed (changes requested):
Thanks for your work! I still found something to nag about and left some comments. What about some reconnection strategy, if the MQTT server goes down? Is this done automatically? 

And can you please open this PR against https://github.com/arnonym/ha-plugins/tree/next? Because this repo is automatically generated from the build.sh script, and I can't merge from here to the main repo. Thanks!

Commit:
fixing comments

eporsche commented:
What about some reconnection strategy, if the MQTT server goes down? Is this done automatically?
 
Unfortunatly I cannot use loop_forever or loop_start from the mqtt library. 
Those would handle automatic reconnects, but they would also push the client out of the main thread.
If that happens the sip library will complain about "not allowing writes from a not registered thread".
Well there is a way to register threads before calling any of the sip functions, but this registry is not type safe - and will present another issue.

In the end i implemented the reconnect in the same way as loop_forever - unfortuantely at least for me - it will hang in a reconnect loop, once the mqtt broker is away.

Are you ok with the changes? If yes - I will open the pull request to the other github repo.

arnonym commented:
One little thing left, the rest looks fine!

I would like to add some information about the stand-alone mode to the README. Do you feel appointed to do so? If not, I will write something and you could do a review of that, because it's not actually my use-case and I might get something wrong.

arnonym commented on ha-sip/src/mqtt.py:
```diff
@@ -0,0 +1,51 @@
+                 username: str,
+                 password: str,
+                 topic: str,
+                 command_handler: callable
```
This should be ```command_handler: CommandHandler```.

If you have a circular import because of that, you can use
```
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from command_handler import CommandHandler
```
to solve that.

arnonym commented on ha-sip/src/main.py:
```diff
@@ -99,7 +105,25 @@ def main():
+                except:
+                    log(None, 'Reconnect to mqtt broker failed. Trying again....')
+                    time.sleep(1)
+            mqtt_client.loop()
```
Maybe another thought: can we move this reconnect logic to mqtt_client.loop() and rename loop to handle? This would clean up these internals from main.py

Commit:
added a readme and fixed a comment

Commit:
language

Commit:
moved reconnect logic to MQTTClient

eporsche commented:
Added the stand-alone mode to the Readme - ready for review.

Commit:
renamed loop to handle

arnonym commented on README.md:
````diff
@@ -523,12 +524,38 @@ I would like to hear from you in which scenario you are using ha-sip!
    { "command": "dial", "number": "sip:**620@fritz.box", "menu": { "message": "Hello from ha-sip.", "language": "en" } }
    ```
+
+## Stand-alone mode
````
very nice, thanks!

arnonym replied:
Can you please remove the white space changes above?

arnonym commented on README.md:
```diff
@@ -523,12 +524,38 @@ I would like to hear from you in which scenario you are using ha-sip!
+1. Follow the instructions from home assistant to set up a working MQTT broker and install the MQTT integration [MQTT Broker](https://www.home-assistant.io/integrations/mqtt/)
+2. Copy `.env.example` to `.env` and replace the variable place-holders with your real configuration.
+3. Make sure you switched the COMMAND_SOURCE in your .env file from "stdin" to "mqtt" and set the BROKER_* variables to connect to your MQTT broker address
+4. Install [docker-compose](https://docs.docker.com/compose/install/) (Hint: Eventhough we just run one service, a simple docker run --env-file won't work, since it's not un-quoting the .env variables correctly)
```
docker-compose is deprecated. It's `docker compose` now. See https://docs.docker.com/compose/migrate/

arnonym commented on ha-sip/src/mqtt.py:
```diff
@@ -0,0 +1,57 @@
+        self.client.connect(self.broker_address, self.port, 60)
+
+    def handle(self):
+        if not self.client.is_connected():
```
That's much better! Thanks!

Commit:
whitespaces

Commit:
language

arnonym commented:
Everything's looking good! Please create a PR to ha-plugins on the next branch!
